File: test-headless/Test/Headless/Lua/UnitAiEncounter.hs
{-# LANGUAGE Strict #-}
-- | Persistent ruin-guard behavior (#916). Runs the shipped encounter AI in
--   a bare Lua VM and pins exact hostility, deterministic target selection,
--   leash/memory disengagement, return-home suppression, and episode notices.
module Test.Headless.Lua.UnitAiEncounter (spec) where

import UPrelude
import Test.Hspec
import qualified HsLua as Lua
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE

runsOk ∷ Text → Expectation
runsOk chunkText = do
    result ← Lua.run @Lua.Exception $ do
        Lua.openlibs
        status ← Lua.dostring (TE.encodeUtf8 chunkText)
        case status of
            Lua.OK → return Nothing
            _ → do
                err ← Lua.tostring (-1)
                return (Just (maybe "<no message>" TE.decodeUtf8Lenient err))
    maybe (pure ()) (expectationFailure . T.unpack) result

lns ∷ [Text] → Text
lns = T.intercalate "\n"

prelude ∷ Text
prelude = lns
    [ "NOW, EVENTS, STOPS, MOVES, CLEARS = 0, 0, 0, 0, 0"
    , "EVENT_MESSAGES, LAST_ATTACKER = {}, nil"
    , "COMMAND, STATES, EPISODES = nil, {}, {}"
    , "VISIBLE = { {x=5,y=5}, {x=7,y=5} }"
    , "IDS = {3,1,2}"
    , "INFO = {"
    , "  [1]={gridX=5,gridY=5,page='world',name='guard'},"
    , "  [2]={gridX=5,gridY=5,page='world',name='near'},"
    , "  [3]={gridX=7,gridY=5,page='world',name='far'} }"
    , "POSE = { [1]='standing', [2]='standing', [3]='standing' }"
    , "REL = { [2]='hostile', [3]='hostile' }"
    , "OCCUPANT = { uid=1, home_x=5, home_y=5,"
    , "  engaged=false, returning=false }"
    , "LOCATION = { instance_id=41, id='ruin_small', name='Old Ruin',"
    , "  lifecycle='discovered', discovered=true,"
    , "  bounds={min_x=0,max_x=10,min_y=0,max_y=10},"
    , "  encounter={cleared=false,activated=false,episode_active=false,"
    , "    aggression_announced=false,disengage_announced=false,"
    , "    occupants={OCCUPANT}} }"
    , "engine = { gameTime=function() return NOW end,"
    , "  emitEventForUnit=function(_,message,uid,x,y) EVENTS=EVENTS+1;"
    , "    table.insert(EVENT_MESSAGES,{message=message,uid=uid,x=x,y=y}) end }"
    , "world = { listPlacedLocations=function() return {LOCATION} end,"
    , "  setLocationEncounterOccupantState=function(i,u,e,r,p)"
    , "    table.insert(STATES,{i=i,u=u,e=e,r=r,p=p}) end,"
    , "  setLocationEncounterEpisodeState=function(i,a,g,d,p)"
    , "    table.insert(EPISODES,{i=i,a=a,g=g,d=d,p=p}) end }"
    , "faction = { relation=function(_,targetFaction)"
    , "  return REL[targetFaction] or 'neutral' end }"
    , "unit = { getInfo=function(uid) return INFO[uid] end,"
    , "  exists=function(uid) return INFO[uid] ~= nil end,"
    , "  getPose=function(uid) return POSE[uid] end,"
    , "  getFaction=function(uid) return uid end,"
    , "  getLastAttacker=function() return LAST_ATTACKER end,"
    , "  getVisibleTiles=function() return VISIBLE end,"
    , "  getAllIds=function() return IDS end,"
    , "  clearAnimOverride=function() CLEARS=CLEARS+1 end,"
    , "  stop=function() STOPS=STOPS+1 end,"
    , "  moveTo=function(_,x,y) MOVES=MOVES+1; LAST_MOVE={x=x,y=y} end }"
    , "REGISTER_OPTIONS=nil"
    , "package.loaded['scripts.unit_ai'] = {"
    , "  commandAttack=function(_,target) COMMAND=target end,"
    , "  setConfig=function() end, registerActions=function(_,_,options)"
    , "    REGISTER_OPTIONS=options end }"
    , "package.loaded['scripts.unit_ai_core'] = {"
    , "  isGoalActive=function(s,name) return s.activeGoal == name end,"
    , "  markGoalAccomplished=function(s,name)"
    , "    if s.activeGoal == name then s.activeGoal=nil end end }"
    , "package.loaded['scripts.unit_ai_page'] = {"
    , "  same=function(a,b) return a == b end }"
    , "package.loaded['scripts.movement_speed'] = { comfort=function() return 1 end }"
    , "package.loaded['scripts.unit_ai_combat_lunge'] = {"
    , "  clear=function() CLEARS=CLEARS+1 end }"
    , "local encounter = require('scripts.unit_ai_encounter')"
    ]

spec ∷ Spec
spec = describe "persistent ruin encounter AI (#916)" $ do
    it "selects the lowest-id visible same-page exact hostile and announces aggression once per encounter episode" $
        runsOk $ lns
            [ prelude
            , "local s = {}"
            , "assert(encounter.engageUtility(1,s) == 8)"
            , "encounter.engageExecute(1,s)"
            , "assert(COMMAND == 2, 'target selection must be deterministic')"
            , "assert(s.ruinEncounterCombat == true)"
            , "assert(EVENTS == 1 and #STATES == 1 and #EPISODES == 1)"
            , "assert(string.find(EVENT_MESSAGES[1].message, 'Old Ruin', 1, true))"
            , "assert(STATES[1].i == 41 and STATES[1].u == 1)"
            , "assert(STATES[1].e and not STATES[1].r)"
            , "assert(EPISODES[1].a and EPISODES[1].g and not EPISODES[1].d)"
            , "encounter.engageExecute(1,s)"
            , "assert(EVENTS == 1, 'aggression repeats within one episode')"
            , "INFO[4]={gridX=6,gridY=5,page='world',name='other guard'}"
            , "POSE[4]='standing'"
            , "local other={uid=4,home_x=6,home_y=5,engaged=false,returning=false}"
            , "table.insert(LOCATION.encounter.occupants,other)"
            , "encounter.engageExecute(4,{})"
            , "assert(EVENTS == 1, 'a second occupant repeated aggression')"
            , "LOCATION.encounter.episode_active=true"
            , "encounter.engageExecute(1,{})"
            , "assert(EVENTS == 1, 'persisted active episode was ignored')"
            ]

    it "keeps pre-discovery aggression and disengagement events private" $
        runsOk $ lns
            [ prelude
            , "LOCATION.lifecycle='unknown'; LOCATION.discovered=false"
            , "local s={}"
            , "encounter.engageExecute(1,s)"
            , "assert(COMMAND == 2 and EVENTS == 0)"
            , "assert(#EPISODES == 1 and EPISODES[1].a"
            , "  and not EPISODES[1].g and not EPISODES[1].d)"
            , "LOCATION.encounter.episode_active=true"
            , "LOCATION.encounter.aggression_announced=false"
            , "OCCUPANT.engaged=true"
            , "INFO[1].gridX=23; s.activeGoal='attack'; s.attackTargetUid=2"
            , "assert(encounter.guardUtility(1,s) == 10)"
            , "encounter.guardExecute(1,s)"
            , "assert(EVENTS == 0, 'hidden disengagement leaked the ruin')"
            , "local episode=EPISODES[#EPISODES]"
            , "assert(not episode.a and not episode.g and not episode.d)"
            ]

    it "rejects neutral, invisible, wrong-page, dead, and cleared targets" $
        runsOk $ lns
            [ prelude
            , "IDS={1,2}; REL[2]='neutral'"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "REL[2]='hostile'; INFO[2].page='other'"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "INFO[2].page='world'; VISIBLE={}"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "VISIBLE={{x=5,y=5}}; POSE[2]='dead'"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "POSE[2]='standing'; LOCATION.encounter.cleared=true"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            ]

    it "retaliates against a fresh same-page hostile hit without requiring sight" $
        runsOk $ lns
            [ prelude
            , "VISIBLE={}; LAST_ATTACKER={uid=2,at=0}; NOW=5"
            , "local s={}"
            , "assert(encounter.engageUtility(1,s) == 8)"
            , "encounter.engageExecute(1,s)"
            , "assert(COMMAND == 2, 'recent attacker was not acquired')"
            , "assert(s.ruinLastSeenAt == 5 and s.ruinLastSeenX == 5"
            , "  and s.ruinLastSeenY == 5)"
            , "assert(s.ruinEncounterCombat == true)"
            , "local function rejected(mut)"
            , "  INFO[2]={gridX=5,gridY=5,page='world',name='near'}"
            , "  POSE[2]='standing'; REL[2]='hostile'; NOW=5"
            , "  LAST_ATTACKER={uid=2,at=0}; mut()"
            , "  assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "end"
            , "rejected(function() REL[2]='neutral' end)"
            , "rejected(function() INFO[2].page='other' end)"
            , "rejected(function() INFO[2].gridX=23 end)"
            , "rejected(function() POSE[2]='dead' end)"
            , "rejected(function() NOW=10.01 end)"
            , "rejected(function() LOCATION.encounter.cleared=true end)"
            ]

    it "uses an inclusive 12-tile Chebyshev leash and suppresses reacquisition until home" $
        runsOk $ lns
            [ prelude
            , "INFO[2].gridX=22; VISIBLE={{x=22,y=5}}"
            , "assert(encounter.engageUtility(1,{}) == 8)"
            , "INFO[2].gridX=22.01; VISIBLE={{x=22,y=5}}"
            , "assert(encounter.engageUtility(1,{}) == -math.huge)"
            , "INFO[2].gridX=5; VISIBLE={{x=5,y=5}}"
            , "INFO[1].gridX=22.01"
            , "LOCATION.encounter.episode_active=true"
            , "LOCATION.encounter.aggression_announced=true"
            , "local s={activeGoal='attack',attackTargetUid=2}"
            , "assert(encounter.guardUtility(1,s) == 10)"
            , "encounter.guardExecute(1,s)"
            , "assert(s.ruinReturning and s.attackTargetUid == nil)"
            , "assert(EVENTS == 1 and MOVES == 1 and CLEARS == 2)"
            , "assert(STATES[1].r and not STATES[1].e)"
            , "assert(#EPISODES == 1 and not EPISODES[1].a"
            , "  and EPISODES[1].g and EPISODES[1].d)"
            , "assert(encounter.engageUtility(1,s) == -math.huge)"
            , "encounter.guardExecute(1,s)"
            , "assert(EVENTS == 1, 'disengage repeats within one episode')"
            , "INFO[1].gridX=5; INFO[1].gridY=5"
            , "encounter.guardExecute(1,s)"
            , "assert(not s.ruinReturning and not s.ruinEncounterCombat)"
            , "assert(STOPS == 2)"
            , "local last=STATES[#STATES]"
            , "assert(not last.e and not last.r and not last.a and not last.d)"
            , "assert(encounter.engageUtility(1,s) == 8)"
            ]

    it "re-enters return-home cleanup from persisted returning state after load" $
        runsOk $ lns
            [ prelude
            , "INFO[1].gridX=9; OCCUPANT.returning=true"
            , "LOCATION.encounter.episode_active=true"
            , "LOCATION.encounter.aggression_announced=true"
            , "local s={activeGoal='attack',attackTargetUid=2}"
            , "assert(encounter.guardUtility(1,s) == 10)"
            , "encounter.guardExecute(1,s)"
            , "assert(s.ruinReturning and s.attackTargetUid == nil)"
            , "assert(MOVES == 1 and CLEARS == 2)"
            ]

    it "pursues only the last seen tile for ten seconds, then disengages \
       \for every invalid target condition" $
        runsOk $ lns
            [ prelude
            , "IDS={1,2}; local s={activeGoal='attack',attackTargetUid=2}"
            , "assert(encounter.guardUtility(1,s) == -math.huge)"
            , "assert(s.ruinLastSeenX == 5 and s.ruinLastSeenY == 5)"
            , "VISIBLE={}; INFO[1].gridX=0; INFO[2].gridX=9; INFO[2].gridY=8; NOW=10"
            , "assert(encounter.guardUtility(1,s) == -math.huge)"
            , "assert(encounter.memoryUtility(1,s) == 8.5)"
            , "encounter.memoryExecute(1,s)"
            , "assert(LAST_MOVE.x == 5 and LAST_MOVE.y == 5,"
            , "  'memory pursuit followed the hidden live target')"
            , "NOW=10.01; assert(encounter.guardUtility(1,s) == 10)"
            , "assert(encounter.memoryUtility(1,s) == -math.huge)"
            , "local function invalid(mut)"
            , "  INFO[2]={gridX=5,gridY=5,page='world',name='near'}"
            , "  POSE[2]='standing'; REL[2]='hostile'; VISIBLE={{x=5,y=5}}"
            , "  local x={activeGoal='attack',attackTargetUid=2,ruinLastSeenAt=NOW}"
            , "  mut(); assert(encounter.guardUtility(1,x) == 10)"
            , "end"
            , "invalid(function() INFO[2]=nil end)"
            , "invalid(function() POSE[2]='dead' end)"
            , "invalid(function() INFO[2].page='other' end)"
            , "invalid(function() REL[2]='friendly' end)"
            ]

    it "keeps generic retaliation excluded after adding encounter-owned hit acquisition" $
        runsOk $ lns
            [ prelude
            , "encounter.register({idleUtility=function() return 0 end,"
            , "  idleExecute=function() end})"
            , "assert(REGISTER_OPTIONS ~= nil)"
            , "assert(REGISTER_OPTIONS.excludeUniversal.engage == true)"
            ]

File: scripts/unit_ai_encounter.lua
-- Persistent ruin-encounter guard behavior (#916).
--
-- Membership, home positions, bounds and clearance live on the placed
-- LocationInstance. This module only owns the decision-time episode state:
-- which visible hostile was last seen, whether the guard is returning, and
-- the exactly-once aggression/disengagement notices for the current episode.

local unitAi = package.loaded["scripts.unit_ai"]
local core = require("scripts.unit_ai_core")
local page = require("scripts.unit_ai_page")
local mv = require("scripts.movement_speed")
local claims = require("scripts.unit_ai_claims")

local M = {}

local LEASH_TILES = 12
local SIGHT_MEMORY_SEC = 10
local RETALIATION_WINDOW_SEC = 10
local RETURN_UTILITY = 10
local MEMORY_UTILITY = 8.5
local ENGAGE_UTILITY = 8

-- World mutations are queued, while several guards can decide in one Lua
-- update. Keep the just-issued episode/participation edges here as a
-- same-process overlay until world.listPlacedLocations reflects them; the
-- persisted encounter remains the save/load authority.
local localEpisodeActive = claims.track({})
local localEpisodeAggression = claims.track({})
local localEpisodeDisengaged = claims.track({})
local localParticipation = claims.track({})

local function encounterKey(binding)
    return tostring(binding.info.page) .. ":"
        .. tostring(binding.location.instance_id)
end

local function participationFor(key)
    local states = localParticipation[key]
    if not states then
        states = {}
        localParticipation[key] = states
    end
    return states
end

local function overlayValue(persisted, overlay, key)
    if overlay[key] ~= nil then return overlay[key] end
    return persisted
end

local function label(info)
    if info and info.name and info.name ~= "" then return info.name end
    return info and info.displayName or "unit"
end

local function locationIsDiscovered(binding)
    return binding.location.discovered == true
end

-- Re-derive the persisted binding every thought tick. This deliberately has
-- no fallback for hand-stamped ruins: they own no encounter instance.
local function bindingFor(uid)
    local info = unit.getInfo(uid)
    if not info or not info.page then return nil end
    for _, location in ipairs(world.listPlacedLocations(info.page) or {}) do
        local encounter = location.encounter
        if encounter then
            for _, occupant in ipairs(encounter.occupants or {}) do
                if occupant.uid == uid then
                    return { location = location, occupant = occupant,
                             info = info }
                end
            end
        end
    end
    return nil
end

local function relationIsHostile(uid, target)
    return faction.relation(unit.getFaction(uid), unit.getFaction(target))
        == "hostile"
end

local function targetVisible(uid, targetInfo)
    if not targetInfo then return false end
    local tx, ty = math.floor(targetInfo.gridX), math.floor(targetInfo.gridY)
    for _, tile in ipairs(unit.getVisibleTiles(uid) or {}) do
        if tile.x == tx and tile.y == ty then return true end
    end
    return false
end

local function outsideLeash(info, bounds)
    return info.gridX < bounds.min_x - LEASH_TILES
        or info.gridX > bounds.max_x + LEASH_TILES
        or info.gridY < bounds.min_y - LEASH_TILES
        or info.gridY > bounds.max_y + LEASH_TILES
end

local function atHome(info, occupant)
    return math.max(math.abs(info.gridX - occupant.home_x),
                    math.abs(info.gridY - occupant.home_y)) <= 0.5
end

local function persistOccupantState(binding, uid, engaged, returning)
    world.setLocationEncounterOccupantState(
        binding.location.instance_id, uid, engaged, returning,
        binding.info.page)
end

local function persistEpisodeState(binding, active, aggressionAnnounced,
                                   disengageAnnounced)
    world.setLocationEncounterEpisodeState(
        binding.location.instance_id, active, aggressionAnnounced,
        disengageAnnounced, binding.info.page)
end

local function clearCombat(uid, s)
    s.attackTargetUid = nil
    s.committed = nil
    s.attackLastMoveTo = nil
    require("scripts.unit_ai_combat_lunge").clear(s)
    core.markGoalAccomplished(s, "attack")
    s.ruinEncounterCombat = nil
    unit.clearAnimOverride(uid)
    unit.stop(uid)
end

local function disengageReason(uid, s, binding)
    local target = s.attackTargetUid
    if not target then return nil end
    if not unit.exists(target) then return "target disappeared" end
    if unit.getPose(target) == "dead" then return "target died" end
    local targetInfo = unit.getInfo(target)
    if not targetInfo or not page.same(binding.info.page, targetInfo.page) then
        return "target left the page"
    end
    if not relationIsHostile(uid, target) then
        return "faction relation is no longer hostile"
    end
    if outsideLeash(targetInfo, binding.location.bounds) then
        return "target crossed the ruin leash"
    end
    if outsideLeash(binding.info, binding.location.bounds) then
        return "guard crossed the ruin leash"
    end
    if targetVisible(uid, targetInfo) then
        s.ruinLastSeenAt = engine.gameTime()
        s.ruinLastSeenX = targetInfo.gridX
        s.ruinLastSeenY = targetInfo.gridY
        return nil
    end
    if engine.gameTime() - (s.ruinLastSeenAt or -math.huge)
       > SIGHT_MEMORY_SEC then
        return "target lost from sight"
    end
    return nil
end

-- While canonical sight is lost, the encounter guard owns pursuit at the
-- last observed tile. Leaving attack_target in charge here would let it read
-- the target's current hidden position every tick, turning ten-second memory
-- into perfect tracking through terrain and facing loss.
local function memoryUtility(uid, s)
    if not s.attackTargetUid then return -math.huge end
    local binding = bindingFor(uid)
    if not binding or disengageReason(uid, s, binding)
       or targetVisible(uid, unit.getInfo(s.attackTargetUid))
       or s.ruinLastSeenX == nil or s.ruinLastSeenY == nil then
        return -math.huge
    end
    return MEMORY_UTILITY
end

local function memoryExecute(uid, s)
    if s.ruinLastSeenX == nil or s.ruinLastSeenY == nil then return end
    local info = unit.getInfo(uid)
    if info and math.max(math.abs(info.gridX - s.ruinLastSeenX),
                         math.abs(info.gridY - s.ruinLastSeenY)) <= 0.5 then
        unit.stop(uid)
        return
    end
    unit.moveTo(uid, s.ruinLastSeenX, s.ruinLastSeenY, mv.comfort(uid))
end

local function guardUtility(uid, s)
    local binding = bindingFor(uid)
    if not binding then return -math.huge end
    if s.ruinReturning or binding.occupant.returning then
        return RETURN_UTILITY
    end
    if s.attackTargetUid then
        local reason = disengageReason(uid, s, binding)
        if reason then
            s.ruinDisengageReason = reason
            return RETURN_UTILITY
        end
    elseif not atHome(binding.info, binding.occupant) then
        s.ruinDisengageReason = "returning to assigned post"
        return RETURN_UTILITY
    end
    return -math.huge
end

local function guardExecute(uid, s)
    local binding = bindingFor(uid)
    if not binding then return end
    local key = encounterKey(binding)
    local participation = participationFor(key)
    if not s.ruinReturning then
        clearCombat(uid, s)
        s.ruinReturning = true
        participation[uid] = "returning"
        persistOccupantState(binding, uid, false, true)
    end

    -- The episode ends only after every surviving participant has broken
    -- contact. Local participation overlays queued writes so two guards
    -- disengaging in the same update still produce exactly one notice.
    local anyEngaged = false
    for _, occupant in ipairs(binding.location.encounter.occupants or {}) do
        if unit.exists(occupant.uid) and unit.getPose(occupant.uid) ~= "dead" then
            local localState = participation[occupant.uid]
            local engaged = localState == "engaged"
                or (localState == nil and occupant.engaged)
            if engaged then anyEngaged = true; break end
        end
    end
    local episodeActive = overlayValue(
        binding.location.encounter.episode_active, localEpisodeActive, key)
    local disengageAnnounced = overlayValue(
        binding.location.encounter.disengage_announced,
        localEpisodeDisengaged, key)
    if episodeActive and not anyEngaged and not disengageAnnounced then
        local visible = locationIsDiscovered(binding)
        if visible then
            engine.emitEventForUnit("unit_event", string.format(
                "%s disengaged at %s (%s)", label(binding.info),
                binding.location.name, s.ruinDisengageReason or "returning"),
                uid, binding.info.gridX, binding.info.gridY)
        end
        local aggressionAnnounced = overlayValue(
            binding.location.encounter.aggression_announced,
            localEpisodeAggression, key)
        localEpisodeActive[key] = false
        localEpisodeDisengaged[key] = visible
        persistEpisodeState(binding, false, aggressionAnnounced, visible)
    end
    if atHome(binding.info, binding.occupant) then
        unit.stop(uid)
        participation[uid] = "home"
        persistOccupantState(binding, uid, false, false)
        s.ruinReturning = nil
        s.ruinLastSeenAt = nil
        s.ruinLastSeenX = nil
        s.ruinLastSeenY = nil
        s.ruinDisengageReason = nil
        return
    end
    unit.moveTo(uid, binding.occupant.home_x, binding.occupant.home_y,
                mv.comfort(uid))
end

local function visibleHostile(uid, binding)
    if binding.location.encounter.cleared then return nil end
    local visible = {}
    for _, tile in ipairs(unit.getVisibleTiles(uid) or {}) do
        visible[tostring(tile.x) .. ":" .. tostring(tile.y)] = true
    end
    local ids = unit.getAllIds() or {}
    table.sort(ids)
    for _, target in ipairs(ids) do
        if target ~= uid and unit.getPose(target) ~= "dead" then
            local info = unit.getInfo(target)
            local key = info and (tostring(math.floor(info.gridX)) .. ":"
                .. tostring(math.floor(info.gridY))) or nil
            if info and page.same(binding.info.page, info.page)
               and visible[key]
               and not outsideLeash(info, binding.location.bounds)
               and relationIsHostile(uid, target) then
                return target
            end
        end
    end
    return nil
end

local function recentHostileAttacker(uid, binding)
    if binding.location.encounter.cleared then return nil end
    local attacker = unit.getLastAttacker(uid)
    if not attacker or attacker.uid == uid
       or engine.gameTime() - (attacker.at or 0) > RETALIATION_WINDOW_SEC
       or not unit.exists(attacker.uid)
       or unit.getPose(attacker.uid) == "dead" then
        return nil
    end
    local info = unit.getInfo(attacker.uid)
    if not info or not page.same(binding.info.page, info.page)
       or outsideLeash(info, binding.location.bounds)
       or not relationIsHostile(uid, attacker.uid) then
        return nil
    end
    return attacker.uid
end

local function hostileTarget(uid, binding)
    local attacker = recentHostileAttacker(uid, binding)
    if attacker then return attacker, false end
    return visibleHostile(uid, binding), true
end

local function engageUtility(uid, s)
    if s.ruinReturning or core.isGoalActive(s, "retreat")
       or core.isGoalActive(s, "attack") then return -math.huge end
    local binding = bindingFor(uid)
    local localState = binding and participationFor(encounterKey(binding))[uid]
    if not binding or binding.occupant.returning or localState == "returning"
       or outsideLeash(binding.info, binding.location.bounds) then
        return -math.huge
    end
    local target = hostileTarget(uid, binding)
    return target and ENGAGE_UTILITY or -math.huge
end

local function engageExecute(uid, s)
    local binding = bindingFor(uid)
    local target, requiresSight = nil, true
    if binding then target, requiresSight = hostileTarget(uid, binding) end
    if not target or not unit.exists(target) then return end
    local targetInfo = unit.getInfo(target)
    if not targetInfo or not page.same(binding.info.page, targetInfo.page)
       or outsideLeash(targetInfo, binding.location.bounds)
       or not relationIsHostile(uid, target)
       or (requiresSight and not targetVisible(uid, targetInfo)) then
        return
    end
    s.ruinLastSeenAt = engine.gameTime()
    s.ruinLastSeenX = targetInfo.gridX
    s.ruinLastSeenY = targetInfo.gridY
    s.ruinEncounterCombat = true
    unitAi.commandAttack(uid, target)
    local key = encounterKey(binding)
    local participation = participationFor(key)
    participation[uid] = "engaged"
    local isNewEpisode = not overlayValue(
        binding.location.encounter.episode_active, localEpisodeActive, key)
    if isNewEpisode then
        local visible = locationIsDiscovered(binding)
        localEpisodeActive[key] = true
        localEpisodeDisengaged[key] = false
        localEpisodeAggression[key] = visible
        if visible then
            engine.emitEventForUnit("unit_event", string.format(
                "%s at %s attacks %s because their faction relation is hostile",
                label(binding.info), binding.location.name, label(targetInfo)),
                uid, binding.info.gridX, binding.info.gridY)
        end
        persistEpisodeState(binding, true, visible, false)
    end
    persistOccupantState(binding, uid, true, false)
end

M.guardUtility = guardUtility
M.guardExecute = guardExecute
M.memoryUtility = memoryUtility
M.memoryExecute = memoryExecute
M.engageUtility = engageUtility
M.engageExecute = engageExecute

function M.register(needs)
    unitAi.setConfig("nomad_primitive", {
        thought_interval = 0.25,
        thought_jitter = 0.05,
        combat_thought_interval = 0.1,
        stuck_walk_timeout = 6.0,
    })
    unitAi.registerActions("nomad_primitive", {
        { name = "ruin_guard", utility = guardUtility,
          execute = guardExecute },
        { name = "ruin_memory", utility = memoryUtility,
          execute = memoryExecute },
        { name = "ruin_engage", utility = engageUtility,
          execute = engageExecute },
        { name = "idle", utility = needs.idleUtility,
          execute = needs.idleExecute },
    }, {
        -- Generic engage reacts to a recent hit without consulting faction,
        -- page, or leash rules. Ruin occupants acquire ONLY through the
        -- encounter-owned visible-hostile and qualified recent-attacker
        -- sources above; retreat and attack execution remain universal once
        -- those rules have selected a target.
        excludeUniversal = { engage = true },
    })
end

return M
